reactor-core/src/main/java/reactor/core/publisher/FluxSwitchOnFirst.java

```diff
@@ -19,610 +19,1029 @@
+				return;
+			}
+
+			this.s.cancel();
+
+			if (shouldCancelInbound) {
+				this.parent.cancelAndError();
+			}
+		}
+
+		@Override
+		public Object scanUnsafe(Attr key) {
+			if (key == Attr.PARENT) {
+				return parent;
+			}
+			if (key == Attr.ACTUAL) {
+				return delegate;
+			}
+			if (key == Attr.RUN_STYLE) {
+				return Attr.RunStyle.SYNC;
+			}
+
+			return null;
+		}
+	}
+
+	static final class SwitchOnFirstConditionalControlSubscriber<T>
+			extends Operators.DeferredSubscription
+			implements InnerOperator<T, T>, ControlSubscriber<T>,
+			           Fuseable.ConditionalSubscriber<T> {
+
+		final AbstractSwitchOnFirstMain<?, T>           parent;
+		final Fuseable.ConditionalSubscriber<? super T> delegate;
+		final boolean                                   cancelSourceOnComplete;
+
+		boolean done;
+
+		SwitchOnFirstConditionalControlSubscriber(AbstractSwitchOnFirstMain<?, T> parent,
+				Fuseable.ConditionalSubscriber<? super T> delegate,
+				boolean cancelSourceOnComplete) {
+			this.parent = parent;
+			this.delegate = delegate;
+			this.cancelSourceOnComplete = cancelSourceOnComplete;
+		}
+
+		@Override
+		public void sendSubscription() {
+			delegate.onSubscribe(this);
+		}
+
+		@Override
+		public void onSubscribe(Subscription s) {
+			if (set(s)) {
+				final long previousState = markOutboundSubscribed(this.parent);
+
+				if (hasOutboundCancelled(previousState)) {
+					s.cancel();
+				}
+			}
+		}
+
+		@Override
+		public CoreSubscriber<? super T> actual() {
+			return this.delegate;
+		}
+
+		@Override
+		public boolean tryOnNext(T t) {
+			if (this.done) {
+				Operators.onNextDropped(t, currentContext());
+				return true;
+			}
+
+			return this.delegate.tryOnNext(t);
+		}
+
+		@Override
+		public void onNext(T t) {
+			if (this.done) {
+				Operators.onNextDropped(t, currentContext());
+				return;
+			}
+
+			this.delegate.onNext(t);
+		}
+
+		@Override
+		public void onError(Throwable throwable) {
+			if (this.done) {
+				Operators.onErrorDropped(throwable, currentContext());
+				return;
+			}
+
+			this.done = true;
+
+			final AbstractSwitchOnFirstMain<?, T> parent = this.parent;
+			long previousState = markOutboundTerminated(parent);
+
+			if (hasOutboundCancelled(previousState) || hasOutboundTerminated(previousState)) {
+				return;
+			}
+
+			if (!hasInboundCancelled(previousState) && !hasInboundTerminated(previousState)) {
+				parent.cancelAndError();
+			}
+
+			this.delegate.onError(throwable);
+		}
+
+		@Override
+		public void onComplete() {
+			if (this.done) {
+				return;
+			}
+
+			this.done = true;
+
+			final AbstractSwitchOnFirstMain<?, T> parent = this.parent;
+			long previousState = markOutboundTerminated(parent);
+
+			if (cancelSourceOnComplete && !hasInboundCancelled(previousState) && !hasInboundTerminated(previousState)) {
+				parent.cancelAndError();
+			}
+
+			this.delegate.onComplete();
+		}
+
+		@Override
+		public void cancel() {
+			REQUESTED.lazySet(this, STATE_CANCELLED);
```

Comment:
this comes from `DeferredSubscription`, right? I think it might be beneficial to readability to explicitly prefix both `REQUESTED` and `STATE_CANCELLED` with `DeferredSubscription.`... (also in the non-conditional variant)

Reply:
done